### ISSUE TITLE
JS Dev Tools

### ISSUE THREAD

**Issue by JayPanoz** [FriendsOfEpub/Blitz#7]
TL;DR: dev tools are lacking, debugging CSS and JS is a nightmare, we could provide blitz’ users with barebones tools in JS.

It’s possible (and actually quite easy) to plug the console, which is already there, into the DOM via JavaScript so, we basically can implement lightweight dev tools inside the EPUB file with a tiny script.

In other words, `console.log()` can be written in the EPUB file. By extension, it’s also possible to get computed styles, print the DOM and shit.

It has little impact on performance and would be designed as a plugin (reflowable/fixed-layout, parameters, API, etc.). 

Also, we could theoretically save logs for later retrieving using localStorage and, why not, export them (mail?).

What we could provide (functions/API): 
- DOM tree;
- RS default styles; 
- element’s styles;
- transparent access to console which works with `console.log`, etc. (JS debugging);
- real viewport;
- offset and shit.

Another lead is https://kenneth.io/blog/2014/12/28/taking-chrome-devtools-outside-the-browser/ but I believe those two things are not exclusive: on the one hand, we could provide lightweight tools (JS) and, on the other hand, more powerful tools (chrome) if needed.

**Comment by JayPanoz** [comment]
Wontfix… in this repo. 

First things first. In hindsight, I’d rather have the CSS framework and the CSS template decoupled – that was a poor decision on my part – so I would consider adding an extra “debug tools” layer more bloat. 

Then, ROI is blurry to say the least. Those dev tools would be a glorified console you would inline in your EPUBs, with predefined methods. But we’d obviously have to create one for reflowable text, and another one for fixed-layout. So that’s quite a significant amount of work. 

That inline console helped me debug some EPUBs tremendously, but that’s just me. Unfortunately, I’m not sure people are comfortable enough with using JavaScript in ebook production. Command-line can already be a huge showstopper, and developers of tooling have to pretty much build GUIs to be useful to the most people possible. Not complaining about this, mind you, just can’t help thinking we’d have to build a tool automating the creation of a debug version from the original EPUB file. So, it should be its own repository to not bloat this one anyway.

I’d be OK putting some work into the init of such a dedicated repository, but I’m definitely not willing to do all the work. So please make your voice heard if you want to help, so that I can reconsider. For the time being, there’s simply no visible incentive to share/polish the debug tools I’ve been using.
